README.md
# bunyan-logger

NOTE that you're reading the README of the version which targets [Koa v2.x](https://github.com/koajs/koa/tree/v2.x), if you want to read about the old 1.X version, switch to Tag 1.3.0.

Flexible log context and request logging middleware
for [koa](http://koajs.com/) using [bunyan](https://github.com/trentm/node-bunyan).

Inspired by [koa-bunyan](https://github.com/ivpusic/koa-bunyan),
[koa-json-logger](https://github.com/rudijs/koa-json-logger),
[bunyan-request](https://github.com/vvo/bunyan-request), and others.

[![Build Status](https://travis-ci.org/koajs/bunyan-logger.svg?branch=master)](https://travis-ci.org/koajs/bunyan-logger)
[![Coverage Status](https://coveralls.io/repos/koajs/bunyan-logger/badge.png)](https://coveralls.io/r/koajs/bunyan-logger)
[![npm](http://img.shields.io/npm/v/koa-bunyan-logger.svg)](https://www.npmjs.org/package/koa-bunyan-logger)

A primary goal of this module is to be as flexible as possible, while
remaining relatively lightweight, leaving advanced functionality and
customization the app.

```js
var koa = require('koa');
var koaBunyanLogger = require('koa-bunyan-logger');

var app = koa();
app.use(koaBunyanLogger());

app.use(function (ctx, next) {
  ctx.log.info('Got a request from %s for %s', ctx.request.ip, ctx.path);
  return next();
});

app.listen(8000);
```

Server:
```
node examples/simple.js | ./node_modules/.bin/bunyan -o short`
```

Client:
```
curl http://localhost:8000/
```

Server output:
```
07:50:14.014Z  INFO koa: Got a request from ::1 for /
```

### Request logging

```js
app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestIdContext());
app.use(koaBunyanLogger.requestLogger());
```

Server:
```
node examples/requests.js | ./node_modules/.bin/bunyan -o short
```

Client:
```
curl -H "X-Request-Id: 1234" http://localhost:8000/
```

Server output:
```
20:19:24.526Z  INFO koa:   --> GET / (req_id=1234)
    GET / HTTP/1.1
    --
    req.header: {
      "user-agent": "curl/7.30.0",
      "host": "localhost:8000",
      "accept": "*/*",
      "x-request-id": "1234"
    }
20:19:24.527Z  INFO koa:   <-- GET / 1ms (req_id=1234, duration=1, res.status=200, res.message=OK)
    GET / HTTP/1.1
    --
    x-powered-by: koa
    content-type: text/plain; charset=utf-8
    content-length: 11
    --
    req.header: {
      "user-agent": "curl/7.30.0",
      "host": "localhost:8000",
      "accept": "*/*",
      "x-request-id": "1234"
    }
```

### Ignoring specific path from logging

It is possible to skip logs from some endpoints with `ignorePath` option.

```js
app.use(koaBunyanLogger.requestLogger({ ignorePath: ['/ping'] }))
```

### Suppressing default error stack traces

To ensure that stack traces from request handling don't get logged in their
raw non-JSON forms, you can disable the app's default error handler:

```js
app.on('error', function () {});
```

## API Reference

### koaBunyanLogger(logger)

Parameters:
- logger: bunyan logger instance or an object to pass to bunyan.createLogger()

#### Examples

Use an existing logger:

```js
var bunyan = require('bunyan');
var koaBunyanLogger = require('koa-bunyan-logger');

var appLogger = bunyan.createLogger({
  name: 'myapp',
  level: 'debug',
  serializers: bunyan.stdSerializers
});

app.use(koaBunyanLogger(appLogger));
```

Shortcut to create a new logger:

```js
var koaBunyanLogger = require('koa-bunyan-logger');

app.use(koaBunyanLogger({
  name: 'myapp',
  level: 'debug'
}));
```

### koaBunyanLogger.requestLogger(opts)

Options:
- durationField: Name of field to store request duration in ms

- levelFn: Function which will be called with (status, err) and should
  return the name of a log level to be used for the response log entry.
  The default function will log status 400-499 as warn, 500+ as error,
  and all other responses as info.

- updateLogFields: Function which will be called with a single argument,
  an object containing the fields (req, res, err) to be logged with the
  request and response message.

  The function has the opportunity to add or remove fields from the object,
  or return a different object to replace the default set of fields.
  The function will be called using the koa 'this' context, once for the
  request and again for the response.

- updateRequestLogFields: Function which will be called with a request
  fields object when logging a request, after processing updateLogFields.

- updateResponseLogFields: Function which will be called with a response
  fields object when logging a response, after processing updateLogFields.
  It also receives a second argument, err, if an error was thrown.

- formatRequestMessage: Function which will be called to generate a log message
  for logging requests. The function will be called in the context of the
  koa 'this' context and passed the request fields object. It should return
  a string.

- formatResponseMessage: Same as formatRequestLog, but for responses.

#### Examples

Basic usage:

```js
app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestLogger());
```

Add custom fields to include in request and response logs:

```js
app.use(koaBunyanLogger.requestLogger({
  // Custom fields for both request and response
  updateLogFields: function (fields) {
    fields.authorized_user = this.user.id;
    fields.client_version = this.request.get('X-Client-Version');
  },

  // Custom fields for response only
  updateResponseLogFields: function (fields, err) {
    if (err) {
      fields.last_db_query = this.db.lastDbQuery();
    }
  }
}));
```

### koaBunyanLogger.requestIdContext(opts)

Get X-Request-Id header, or if the header does not exist, generates a random
unique id for each request.

Options:
- header: name of header to get request id from
- prop: property to store on context; defaults to 'reqId' e.g. this.reqId
- requestProp: property to store on request; defaults to 'reqId' e.g. this.request.reqId
- field: field to add to log messages in downstream middleware and handlers;
  defaults to 'req_id'

#### Examples

```js
var koaBunyanLogger = require('koa-bunyan-logger');

app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestIdContext());
```

Or use a different header:

```js
var koaBunyanLogger = require('koa-bunyan-logger');

app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestIdContext({
  header: 'Request-Id'
}));
```

By default, the request id will be accessible as this.reqId and this.request.reqId:

```js
var koaBunyanLogger = require('koa-bunyan-logger');

app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestIdContext());

app.use(function (ctx) {
  ctx.response.set('X-Server-Request-Id', ctx.reqId);
  ctx.body = "Hello world";
});
```

### koaBunyanLogger.timeContext(opts)

Adds `time(label)` and `timeEnd(label)` methods to the koa
context, which records the time between the time() and
timeEnd() calls for a given label.

Calls to time() and timeEnd() can be nested or interleaved
as long as they're balanced for each label.

Options:
- logLevel: name of log level to use; defaults to 'trace'
- updateLogFields: function which will be called with
    arguments (fields) in koa context; can update fields or
    return a new object.

#### Examples

```js
var koaBunyanLogger = require('koa-bunyan-logger');

app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestIdContext());
app.use(koaBunyanLogger.timeContext());

app.use(function (ctx) {
  ctx.time('get data');

  return getUser()
  .then(u => {
    return getFriend(u)
    .then(f => [u, f]);
  })
  .then(data => {
    let user = data[0];
    let friends = data[1];

    ctx.timeEnd('get data');
    ctx.time('serialize');
    ctx.body = serialize(user, friends);
    ctx.timeEnd('serialize');
  });
});
```

The same but using async functions

```js
var koaBunyanLogger = require('koa-bunyan-logger');

app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestIdContext());
app.use(koaBunyanLogger.timeContext());

app.use(async function (ctx) {
  ctx.time('get data');
  let user = await getUser();
  let friends = await getFriend(user);
  ctx.timeEnd('get data');

  ctx.time('serialize');
  ctx.body = serialize(user, friends);
  ctx.timeEnd('serialize');
});
```

Example output:
```json
{"name":"koa","hostname":"localhost","pid":9228,"level":10,"label":"get data","duration":102,"msg":"","time":"2014-11-07T01:45:53.711Z","v":0}
{"name":"koa","hostname":"localhost","pid":9228,"level":10,"label":"serialize","duration":401,"msg":"","time":"2014-11-07T01:45:54.116Z","v":0}
```

To return different fields, such as nesting the data under
a single field, add a `updateLogFields` function to the options:

```js
var koaBunyanLogger = require('koa-bunyan-logger');

app.use(koaBunyanLogger());
app.use(koaBunyanLogger.requestIdContext());
app.use(koaBunyanLogger.timeContext({
  updateLogFields: function (fields) {
    return {
      request_trace: {
        name: fields.label,
        time: fields.duration
      }
    };
  }
}));
```

### bunyan export

The internal copy of bunyan is exported as `.bunyan`:

```js
var koaBunyanLogger = require('koa-bunyan-logger');
var bunyan = koaBunyanLogger.bunyan;
```

## Sponsored by

[Pebble Technology!](https://www.pebble.com)

## License

[MIT](https://github.com/koajs/bunyan-logger/blob/master/LICENSE)


History.md

2.1.0 / 2018-10-04
==================

  * ci: Add eslint-config, use only .travis.yml
  * style: Fix linter problems
  * update package-lock
  * refactor: Use async, const-let style, remove generators
  * chore: Update deps
  * chore: Fix vulnerabilities with npm audit
  * chore: Add package-lock.json
  * feat: Add ignorePath options
  * Merge pull request #25 from koajs/2x

2.0.0 / 2017-04-14
==================

Thanks to [ifraixedes](https://github.com/ifraixedes) for
updating and refactoring to work with koa 2.x

  * Merge pull request #16 from venables/update-uuid-2x
  * Merge pull request #15 from marcbachmann/update-uuid
  * chore(package): update uuid to version 3.0.0
  * Include stdSerializers in custom bunyan logger example
  * Merge pull request #12 from umayr/2x
  * Update examples
  * Merge pull request #11 from kevinawoo/patch-1
  * fixed example typo.
  * Merge pull request #10 from alxarch/patch-1
  * Fix wrong method name

1.3.0 / 2015-10-26
==================

  * Bump bunyan to 1.5.0, bump test dependencies
  * Export bunyan
  * Fix req object (was incorrectly using koa's request object)

1.2.0 / 2015-01-23
==================

  * Merge pull request #5 from pebble/standard-req-res
  * Pass this.req/this.res instead of this.request/this.response

1.1.1 / 2015-01-13
==================

 * use named middleware for better debugging #4 [mpal9000](https://github.com/mpal9000)

1.1.0 / 2014-11-10
==================

 * Add timeContext() utility to enable this.time() and this.timeEnd()

1.0.1 / 2014-11-03
==================

Fix locking down dependencies. Avoid using bunyan-1.2 since dtrace-provider
has problems on OSX. Move mocha to devDependencies.


index.js
'use strict';

var bunyan = require('bunyan');
var uuid = require('uuid');
var util = require('util');
var onFinished = require('on-finished');

/*
 * If logger is a bunyan logger instance, return it;
 * otherwise, create a new logger with some reasonable defaults.
 */
function createOrUseLogger(logger) {
  if (!logger || !logger.info || !logger.child) {
    var loggerOpts = logger || {};
    loggerOpts.name = loggerOpts.name || 'koa';
    loggerOpts.serializers = loggerOpts.serializers || bunyan.stdSerializers;

    logger = bunyan.createLogger(loggerOpts);
  }

  return logger;
}

/*
 * Koa middleware that adds this.log property to the koa context
 * containing a bunyan logger instance.
 *
 * Parameters:
 *  - loggerInstance: bunyan logger instance, or an object with properties
 *                    that will be passed to bunyan.createLogger. If not
 *                    specified, a default logger will be used.
 */
module.exports = function (loggerInstance) {
  loggerInstance = createOrUseLogger(loggerInstance);

  return function logger(ctx, next) {
    ctx.log = loggerInstance;

    return next();
  };
};

/*
 * Koa middleware that gets a unique request id from a header or
 * generates a new one, and adds the requestId to all messages logged
 * using this.log in downstream middleware and handlers.
 *
 * Must use(koaBunyanLogger()) before using this middleware.
 *
 * Parameters:
 *  - opts: object with optional properties:
 *    - header: name of header to get request id from (default X-Request-Id)
 *    - prop: property to store on 'this' context (default 'reqId')
 *    - requestProp: property to store on 'this.request' (default 'reqId')
 *    - field: log field name for bunyan (default 'req_id')
 */
module.exports.requestIdContext = function (opts) {
  opts = opts || {};

  var header = opts.header || 'X-Request-Id';
  var ctxProp = opts.prop || 'reqId';
  var requestProp = opts.requestProp || 'reqId';
  var logField = opts.field || 'req_id';

  return function requestIdContext(ctx, next) {
    var reqId = ctx.request.get(header) || uuid.v4();

    ctx[ctxProp] = reqId;
    ctx.request[requestProp] = reqId;

    var logFields = {};
    logFields[logField] = reqId;

    if (!ctx.log) {
      throw new Error('must use(koaBunyanLogger()) before this middleware');
    }

    ctx.log = ctx.log.child(logFields);

    return next();
  };
};

/*
 * Logs requests and responses.
 *
 * Must use(koaBunyanLogger()) before using this middleware.
 *
 * Parameters:
 *  - opts: object with optional properties
 *    - durationField: name of duration field
 *    - levelFn: function (status, err)
 *    - updateLogFields: function (data)
 *    - updateRequestLogFields: function (requestData)
 *    - updateResponseLogFields: function (responseData)
 *    - formatRequestMessage: function (requestData)
 *    - formatResponseMessage: function (responseData)
 */
module.exports.requestLogger = function (opts) {
  opts = opts || {};

  var levelFn = opts.levelFn || function (status, err) {
    if (status >= 500) {
      return 'error';
    } else if (status >= 400) {
      return 'warn';
    } else {
      return 'info';
    }
  };

  var durationField = opts.durationField || 'duration';

  var formatRequestMessage = opts.formatRequestMessage || function (data) {
    return util.format('  <-- %s %s',
                       this.request.method, this.request.originalUrl);
  };

  var formatResponseMessage = opts.formatResponseMessage || function (data) {
    return util.format('  --> %s %s %d %sms',
                       this.request.method, this.request.originalUrl,
                       this.status, data[durationField]);
  };

  return function requestLogger(ctx, next) {
    if (Array.isArray(opts.ignorePath) && opts.ignorePath.includes(ctx.path)) {
      return next();
    }

    var requestData = {
      req: ctx.req
    };

    requestData = updateFields(ctx, opts.updateLogFields, requestData);
    requestData = updateFields(ctx, opts.updateRequestLogFields, requestData);

    ctx.log.info(requestData, formatRequestMessage.call(ctx, requestData));

    var startTime = new Date().getTime();
    var err;

    var onResponseFinished = function () {
      var responseData = {
        req: ctx.req,
        res: ctx.res
      };

      if (err) {
        responseData.err = err;
      }

      responseData[durationField] = new Date().getTime() - startTime;

      responseData = updateFields(ctx, opts.updateLogFields, responseData);
      responseData = updateFields(ctx, opts.updateResponseLogFields,
                                  responseData, err);

      var level = levelFn.call(ctx, ctx.status, err);

      ctx.log[level](responseData,
                      formatResponseMessage.call(ctx, responseData));

      // Remove log object to mitigate accidental leaks
      ctx.log = null;
    };

    return next().catch(e => {
      err = e;
    }).then(() => { // Emulate a finally
      // Handle response logging and cleanup when request is finished
      // This ensures that the default error handler is done
      onFinished(ctx.response.res, onResponseFinished.bind(ctx));

      if (err) {
        throw err; // rethrow
      }
    });
  };
};

function updateFields (ctx, func, data, err) {
  if (!func) return data;

  try {
    if (err) {
      return func.call(ctx, data, err) || data;
    } else {
      return func.call(ctx, data) || data;
    }
  } catch (e) {
    ctx.log.error(e);
    return data;
  }
}

/**
 * Middleware which adds methods this.time(label) and this.timeEnd(label)
 * to koa context.
 *
 * Parameters:
 * - opts: object with the following optional properties
 *   - logLevel: name of log level to use; defaults to 'trace'
 *   - updateLogFields: function which will be called with
 *     arguments (fields) in koa context; can update fields or
 *     return a new object.
 *
 * Must use(koaBunyanLogger()) before using this middleware.
 */
module.exports.timeContext = function (opts) {
  opts = opts || {};

  var logLevel = opts.logLevel || 'trace';
  var updateLogFields = opts.updateLogFields;

  return function timeContext(ctx, next) {
    ctx._timeContextStartTimes = {};

    ctx.time = time;
    ctx.timeEnd = timeEnd;

    return next();
  };

  function time (label) {
    /*jshint validthis:true */
    var startTimes = this._timeContextStartTimes;

    if (startTimes[label]) {
      this.log.warn('time() called for previously used label %s', label);
    }

    startTimes[label] = new Date().getTime();
  }

  function timeEnd (label) {
    /*jshint validthis:true */
    var startTimes = this._timeContextStartTimes;
    var startTime = startTimes[label];

    if (!startTime) { // whoops!
      this.log.warn('timeEnd() called without time() for label %s', label);
      return;
    }

    var duration = new Date().getTime() - startTime;
    var fields = {
      label: label,
      duration: duration,
      msg: label + ': ' + duration + 'ms'
    };

    fields = updateFields(this, updateLogFields, fields);
    this.log[logLevel](fields);

    startTimes[label] = null;
  }
};

// Export our copy of bunyan
module.exports.bunyan = bunyan;
